File: 1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.cpp
class Solution {
public:
    void floydWarshall(vector<vector<int>> &v, int INF) {
        int V = v.size();

        for (int k = 0; k < V; k++) {
            for (int i = 0; i < V; i++) {
                for (int j = 0; j < V; j++) {
                    if (v[i][k] != INF && v[k][j] != INF)
                        v[i][j] = min(v[i][j], v[i][k] + v[k][j]);
                }
            }
        }
    }

    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        int INF = 100000000;
        vector<vector<int>> v(n, vector<int>(n, INF));
        for (int i = 0; i < n; i++) v[i][i] = 0;

        for (auto& i : edges) {
            int a = i[0], b = i[1], c = i[2];
            v[a][b] = c;
            v[b][a] = c;
        }

        floydWarshall(v, INF);

        int x = INT_MAX;
        int c = -1;

        for (int i = 0; i < n; i++) {
            int p = 0;
            for (int j = 0; j < n; j++) {
                if (v[i][j] <= distanceThreshold)
                    p++;
            }

            if (p <= x) {
                x = p;
                c = i;
            }
        }

        return c;
    }
};

File: 1456-find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance/find-the-city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.java
class Solution {
    private void helper(ArrayList<ArrayList<int[]>> adj,int[] dist,int i)
    {
         PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
         dist[i] = 0;
        pq.offer(new int[]{0, i});

        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            int d = top[0];
            int u = top[1];

            if (d > dist[u])
                continue;

            for (int p[] : adj.get(u)) {
                int v = p[0];
                int w = p[1];

                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    pq.offer(new int[]{dist[v], v});
                }
            }
        }

    }
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        ArrayList<ArrayList<int[]>> adj = new ArrayList<>();
        for(int i=0;i<n;i++)
        {
            adj.add(new ArrayList<>());
        }
        for(int i=0;i<edges.length;i++)
        {
            int a=edges[i][0];
            int b=edges[i][1];
            int c=edges[i][2];
            adj.get(a).add(new int[]{b,c});
            adj.get(b).add(new int[]{a,c});
        }
        int node=Integer.MIN_VALUE,c=Integer.MAX_VALUE;
        for(int i=0;i<n;i++)
        {
            int[] dist=new int[n];
            int p=0;
            Arrays.fill(dist,Integer.MAX_VALUE);
            helper(adj,dist,i);
            for(int j=0;j<n;j++)
            {
                if(i!=j && dist[j]<=distanceThreshold)
                {
                    p++;
                }
            }
            if(c>=p)
            {
                c=Math.min(c,p);
                node=Math.max(node,i);
            }
        }
        return node;
    }
}